Feature/report refactor

AnOctopus opened the pull request:
Still needs some threading documentation (and maybe explicitly guard against user scripts violating threading requirements?), AppSession constructor cleanup, and a pile of misc report references in variable/parameter names and documentation.

AnOctopus committed:
Rename ReportSession->AppSession, ReportContext->ScriptRunContext

AnOctopus committed:
Rename `add_report_ctx`->`add_script_run_ctx`

AnOctopus committed:
Fix config test failures from null `message_enqueued_callback`

AnOctopus committed:
Rename `REPORT_CONTEXT`->`SCRIPT_RUN_CONTEXT`

AnOctopus committed:
report_thread refactor

Rename module to `script_run_context`
Move `ScriptThread` to separate `script_thread` module

AnOctopus committed:
Rename ReportQueue->ForwardMsgQueue

AnOctopus committed:
Rename AppSessionState enum constants

AnOctopus committed:
Rename Report->SessionData

File name not yet changed because it caused seemingly unrelated test failures

AnOctopus committed:
Rename `report_test` file

This required fixing some persistent side effects from one of the server
tests changing the port in the config.

AnOctopus committed:
Attrs classes for queues and context

AnOctopus committed:
Extract `get_url` class method to module

AnOctopus committed:
Make mypy happy

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

AnOctopus committed:
Replace StringSet with regular set

If the context is only ever accessed from the script thread, its
attributes don't need to be thread safe, and we can avoid this
complexity.

The API for regular sets is a little worse, but switching to something
more standard and familiar is overall a good trade.

AnOctopus committed:
Remove everything related to report sharing (#4124)

Thiago mentioned that this feature is only used by the docs site, and
could be freely removed.

AnOctopus committed:
Extract SessionData construction from AppSession

AnOctopus committed:
Fix AppSession construction in server

AnOctopus committed:
Add threading note to script_thread module

AnOctopus committed:
Clean up ScriptRunner references to "report"

AnOctopus committed:
Fix ScriptThread documentation

AnOctopus committed:
Replace ScriptRunContext refernces to "report"

AnOctopus committed:
Clean up SessionData documentation

AnOctopus committed:
Replace AppSession references to "report" in comments and variables

AnOctopus commented on lib/streamlit/script_thread.py:
```diff
@@ -0,0 +1,68 @@
+from streamlit.state.session_state import SessionState
+from streamlit.uploaded_file_manager import UploadedFileManager
+
+# Note [Threading]
+# There are two kinds of threads in Streamlit, the main thread and script threads.
+# The main thread is started by invoking the Streamlit CLI, and bootstraps the
+# framework and runs the Tornado webserver.
+# A script thread is created by a ScriptRunner when it starts. The script thread
+# is where the ScriptRunner executes, including running the user script itself,
+# processing messages to/from the frontend, and all the Streamlit library function
+# calls in the user script.
+# It is possible for the user script to spawn its own threads, which could call
+# Streamlit functions. We restrict the ScriptRunner's execution control to the
+# script thread. Calling Streamlit functions from other threads is unlikely to
+# work correctly due to lack of ScriptRunContext, so we may add a guard against
+# it in the future.
```
What do you think of this style of documentation?

I wanted to try it out, because it is used throughout GHC (haskell compiler) source code and I think it is good for providing explanation of higher level concepts and concerns.

[Here is a small post about it](https://www.stackbuilders.com/news/the-notes-of-ghc)

tconkling replied:
I love high-level "concept" documentation! I think the standard way to this is similar to other docstrings: use a multiline string (and I think technically it should go above the imports, but after the annoying LICENSE header):

```python
# Blah blah blah license and copyright stuff

"""
Note [Threading]:
There are two kinds of threads in Streamlit...
...
"""

import xyz
```

https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring

AnOctopus committed:
Revert "Remove everything related to report sharing (#4124)"

This reverts commit d7834ba995bd894f82967783a61c4e8406072764.

AnOctopus committed:
Add missing type import for SessionData

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

AnOctopus committed:
Fix AppSession creation in test

AnOctopus committed:
Rename "report" references in protobuf files

AnOctopus committed:
Rename many report references in TS code

AnOctopus committed:
Rename frontend "report" classes

AnOctopus committed:
Rename more frontend "report" references

AnOctopus commented:
@tconkling 
The PR consists of the following kinds of changes:

1. Renaming classes/functions, to replace the concept of "report" with the notion of a script run, an app, or an app session, as appropriate
2. Make some classes use attrs to generate reasonable init, repr, and equality methods (and use slots for space savings)
3. Split out the `ScriptRunContext` from the `ScriptThread` module
4. Add type annotations for a few functions/classes
5. Some higher level documentation on threading, in an experimental format

Separate from this will be PRs for:
1. Excising the unused "report sharing" feature (already reviewed, but split back out to make this PR cleaner)
2. DI of `LocalSourcesWatcher` for `AppSession`, and associated test fixup (#4167)
Which can either be merged in fully separately, or we can get this PR reviewed and then review and merge them in before going to develop

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

tconkling commented:
Notes while reviewing:
- "Report" is renamed "App" everywhere

tconkling commented on frontend/src/lib/AppNode.test.ts:
```diff
@@ -827,18 +827,18 @@ describe("ReportRoot.applyDelta", () => {
-    expect(newRoot.main.getIn([1, 1])?.reportId).toBe("new_report_id")
-    expect(newRoot.sidebar.reportId).toBe(NO_REPORT_ID)
+    expect(newRoot.main.sessionId).toBe("new_app_id")
+    expect(newRoot.main.getIn([0])?.sessionId).toBe(NO_REPORT_ID)
```
NO_REPORT_ID -> NO_SESSION_ID?

tconkling commented on frontend/src/lib/MetricsManager.ts:
```diff
@@ -171,20 +171,20 @@ export class MetricsManager {
-  public setReportHash = (reportHash: string): void => {
-    this.reportHash = reportHash
+  // so they will not include the appHash.
+  public setReportHash = (appHash: string): void => {
```
setReportHash -> setAppHash ?

tconkling commented on frontend/src/lib/MetricsManager.ts:
```diff
@@ -171,20 +171,20 @@ export class MetricsManager {
       ...this.getHostTrackingData(),
       ...MetricsManager.getInstallationData(),
-      reportHash: this.reportHash,
+      appHash: this.appHash,
```
This is a schema change to our metrics events, so we should not do it except at the request of the data team. So this should be changed to `reportHash: this.appHash` (we can name the data appHash internally, just not in the data we send out). 

Also worth adding a comment here explaining that we're using "reportHash" because doing otherwise would break an old schema.

tconkling commented on frontend/src/lib/ScriptRunState.ts:
```diff
@@ -0,0 +1,24 @@
+  RUNNING = "running",
+  RERUN_REQUESTED = "rerunRequested", // report *not* running, but user requested it be re-run
+  STOP_REQUESTED = "stopRequested", // report *is* running, but user requested it be stopped
+  COMPILATION_ERROR = "compilationError", // report's script failed with a compilation error
```
A few stray "report" mentions in the comments here.

tconkling commented on frontend/src/lib/s3helper.ts:
```diff
@@ -20,11 +20,11 @@ import { FETCH_PARAMS } from "./baseconsts"
 
 /**
  * Returns the local path for a static report resource.
- * @param reportId the report ID of the resource to fetch
+ * @param sessionId the report ID of the resource to fetch
  * @param objName the name of the resource to fetch
  */
```
I think this whole class can go away, right? (As part of your static sharing prune.)

AnOctopus replied:
Yep, it is removed in that other PR

tconkling commented on lib/streamlit/app_session.py:
```diff
@@ -630,8 +634,10 @@ def progress(percent):
+        )
 
         if config.get_option("server.liveSave"):
             url_util.print_url("Saved running app", url)
```
This and the next function should probably go away? (Obv, ignore if they're already part of your other PR to remove sharing.)

tconkling commented on lib/streamlit/elements/form.py:
```diff
@@ -179,10 +179,12 @@ def form(self, key: str, clear_on_submit: bool = False):
         if ctx is not None:
-            added_form_id = ctx.form_ids_this_run.add(form_id)
-            if not added_form_id:
+            new_form_id = form_id not in ctx.form_ids_this_run
+            if new_form_id:
+                ctx.form_ids_this_run.add(form_id)
+            else:
```
Why this change? This is very minor, but I think it's basically doubling the work that the function has to do: there are now two set lookups whenever adding to `form_ids_this_run`, whereas before there was just one. (Also, minor nit: `new_form_id` -> `is_new_form_id` - it's a predicate, not a form_id itself.)

AnOctopus replied:
I removed `_StringSet` and replaced it with python's builtin `Set`, but the latter has an API that doesn't tell us if the added value was already in it

tconkling replied:
Ah, makes sense! Carry on then :)

tconkling commented on lib/streamlit/script_run_context.py:
```diff
@@ -0,0 +1,144 @@
+        # via `streamlit run`. Otherwise, the user is likely running a
+        # script "bare", and doesn't need to be warned about streamlit
+        # bits that are irrelevant when not connected to a session.
+        LOGGER.warning("Thread '%s': missing ScriptRunContext" % thread.name)
```
Nit: should use a comma, not the `%` operator: 

```
LOGGER.warning("Thread '%s': missing ScriptRunContext" % thread.name)
``` 

(or use an f-string if we want to do eager string interpolation, but we probably shouldn't usually do that for log strings.)

tconkling commented on lib/streamlit/script_runner.py:
```diff
@@ -73,28 +78,28 @@ def __init__(
         Parameters
         ----------
         session_id : str
-            The ReportSession's id.
+            The AppSession's id.
```
This is the "real" `session_id` (and it's different from the ReportID)

tconkling commented on lib/streamlit/script_thread.py:
```diff
@@ -0,0 +1,65 @@
+# script thread. Calling Streamlit functions from other threads is unlikely to
+# work correctly due to lack of ScriptRunContext, so we may add a guard against
+# it in the future.
+class ScriptThread(threading.Thread):
```
I don't have a strong opinion, but I think we should consider also jettisoning `ScriptThread`. All it does is attach the `ScriptRunContext` to itself - exactly what the `add_script_run_context` already does. There's nothing that makes a `ScriptThread` special or different from any other thread that has the context, so this file is basically a lot of boilerplate.

AnOctopus replied:
That is a solid point; it would be easy to remove.

tconkling commented on lib/streamlit/state/widgets.py:
```diff
@@ -143,8 +143,10 @@ def register_widget(
     # already been registered.
-    added = ctx.widget_ids_this_run.add(widget_id)
-    if not added:
+    new_widget = widget_id not in ctx.widget_ids_this_run
+    if new_widget:
+        ctx.widget_ids_this_run.add(widget_id)
+    else:
```
This is also now a double set-membership check, whereas pre-change there was just a single lookup. What's the purpose? (IMO, it's also less readable than it was before.)

tconkling commented on proto/streamlit/proto/Delta.proto:
```diff
@@ -24,10 +24,10 @@ import "streamlit/proto/ArrowNamedDataSet.proto";
 message Delta {
   oneof type {
-    // Append a new element to the report.
+    // Append a new element to the web app.
```
Micronit: we generally just refer to this as "the frontend"

tconkling commented on proto/streamlit/proto/NewApp.proto:
```diff
@@ -18,21 +18,21 @@ syntax = "proto3";
-// This is the first message that is sent when a new report is run.
-message NewReport {
+// This is the first message that is sent when a new session starts.
+message NewApp {
```
Seems like this should perhaps be named "NewSession", instead of "NewApp"? (Since yeah ,it's fired off every time a session connects to a given app. The app isn't "new" when a session connects.)

tconkling commented on frontend/src/App.tsx:
```diff
@@ -115,20 +115,20 @@ export interface Props {
-  reportName: string
-  reportHash: string | null
-  reportRunState: ReportRunState
+  sessionId: string
```
`sessionId` now has two meanings (we already have a session_id that's used in `ReportSession`). I _think_ this sessionId is just a renamed `Report.report_id`? Which is not really a session ID - it's an ID that's generated per `streamlit run myapp.py` execution, but not per-connected user.

AnOctopus replied:
That is a good point, I missed that there were two meanings.

tconkling reviewed (changes requested):
I think this looks mostly good! There are a few nits, and one easily-fixed "we should not ship this" (a metrics schema change). 

There are a couple naming bits that stand out as needing a tiny bit more discussion:
- We now have two "session_ids" - the one we had before, and the renamed `Report.report_id`.
- "Report" has mostly been renamed to "App" - which makes sense - but there are a few places where it's been renamed to "Script". Is there a guideline about when you think we should use one instead of the other?

Once this stuff is worked out, we should summarize the changes and run it by #eng-core before merging, to make sure they're not caught off guard.

AnOctopus committed:
Rename `NO_REPORT_ID`->`NO_SESSION_ID`

AnOctopus committed:
Fix metrics schema

AnOctopus committed:
Make log string formatting lazy

AnOctopus committed:
NewApp->NewSession

AnOctopus committed:
Update Delta proto comments

AnOctopus committed:
setReportHash->setAppHash

AnOctopus committed:
Update "report" mentions in ScriptRunState comment

AnOctopus committed:
Remove ScriptThread

AnOctopus committed:
Rename sessionId (prev ReportId) to scriptRunId

This removes name collision with the actual session id, and better
reflects the scope of action the id refers to.

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

AnOctopus commented:
@tconkling (can't reply inline to outdated comment I guess?)
I like the idea of switching the note to a string, but I'm unsure about making it a module docstring, since there may be multiple in a module, and they may apply more to a few classes/functions in a module rather than the whole thing

tconkling commented:
IMO it's fine to have as a module docstring even with those caveats, but I also don't feel that strongly either way!

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

AnOctopus committed:
Convert threading note to string

AnOctopus committed:
Fix names in Block test

Introduced by merging in develop

AnOctopus committed:
Fix Block function argument name

AnOctopus committed:
Fix AppNode id naming (session->scriptRun)

AnOctopus committed:
Extract LocalSourcesWatcher to argument for AppSession (#4167)

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

vdonato commented on e2e/specs/st_radio.spec.js:
```diff
@@ -49,9 +49,9 @@ describe("st.radio", () => {
-        "value 4: female"
+          "value 2: female" +
+          "value 3: None" +
+          "value 4: female"
```
mostly inconsequential, but pulling in the latest `develop` changes will get rid of these autoformatter-related diffs.

vdonato commented on frontend/src/App.test.tsx:
```diff
@@ -360,13 +360,13 @@ describe("App.handleNewReport", () => {
     },
   }
-  const NEW_REPORT = new NewReport(NEW_REPORT_JSON)
+  const NEW_REPORT = new NewSession(NEW_REPORT_JSON)
```
nit: looks like the `NEW_REPORT` and `NEW_REPORT_JSON` consts snuck by without being renamed

vdonato commented on lib/streamlit/app_session.py:
```diff
@@ -43,41 +46,41 @@
             The Tornado IOLoop that we're running within.
 
         script_path : str
-            Path of the Python file from which this report is generated.
+            Path of the Python file from which this app is generated.
```
nit: this docstring parameter is now stale with `session_data` replacing `script_path`

vdonato commented on lib/tests/streamlit/session_data_test.py:
```diff
@@ -134,9 +137,10 @@ def test_get_url(self, baseUrl, port, expected_url):
         if port:
             options["server.port"] = port
 
+        # breakpoint()
```
nit: should remove this

vdonato reviewed (approved):
Overall LGTM!

I think there are a few places where the word `report` may have slipped by according to a quick search running `$ grep -nri report` on both `lib/streamlit` and `frontend/src`. It's hard to tell the exact number since many of these will vanish when the report sharing PR is merged, and a few of the instances that the search turns up are false positives (from the "report a bug" button, etc.).

The places that do look like legitimate spots where `report*` instances slipped by include some comments, tests, and places in `forward_message_cache.py`/`ForwardMessageCache.ts` (such as `report_run_count`/`reportRunCount`). 

It'd probably be worthwhile to make one last pass for the word `report` once the sharing PR is merged (and when that pass is made, I wouldn't be opposed to the stragglers being removed in a separate PR since this one is long enough to be difficult to read through as it is).

AnOctopus committed:
Merge branch 'develop' into feature/report-refactor

AnOctopus committed:
Rename NEW_REPORT->NEW_SESSION in App tests

AnOctopus committed:
Update AppSession docstring parameters

AnOctopus committed:
Delete commented out breakpoint

kajarenc committed:
Merge branch 'develop' into feature/report-refactor

kajarenc committed:
Remove empty ReportRunState.ts to avoid frontend build crush
Use `script_run_context` instead of obsolete `get_script_run_ctx` in camera_input.py

tconkling reviewed (approved):
Great! 

(Minor: it looks like `ReportRunState.ts` is an empty file and should be removed?)

richlysakowski commented:
ModuleNotFoundError: No module named 'streamlit.report_thread'

It looks like this was closed and yet I still get the error message above???

What is the fix for this?  I need a solution right away.  

I will post this issue on StackOverflow.com so that it does not linger any longer.  

Please specify everything that must be changed.

tconkling commented:
Hi @richlysakowski - `streamlit.report_thread` was an internal API. We don't recommend that devs use internal APIs, because they're subject to change without notice, and we don't provide support for them. (In this case, ReportThread no longer exists.) 

If you're looking to create additional threads within your Streamlit app, you can do this with the built-in Python thread class, but note that it is _not safe_ to call streamlit APIs from other threads. (If you want to issue Streamlit commands from another thread, you could e.g. have that thread enqueue requests to a shared queue that your script's thread processes.)

richlysakowski commented:
Hi #tconkling, 

Thank you for your fast response.   Streamlit is truly great and a game changer for analytics and app dev in general.

I downloaded an application that uses the module `streamlit.report_thread` (see Streamlit main code below) but I don't know how to correct or change the reference to the `report_thread` object.

I read your feedback above and I will try to remove it from the code, but I think that will not fix it by itself.  

How do I replace this Streamlit thread library with the equivalent Python thread library?   I don't know much about using multi-threading libraries.  I could try to look in the old Streamlit code that had `streamlit.report_thread` module.  I have been successful modifying or stubbing out legacy code, if that is all that is needed.   Any help or guidance you can provide will be much appreciated.  

I searched the code in all project files in PyCharm and the report_thread module **is only referenced ONCE** in the StringIO write function below.  It should be pretty easy to remove or replace, because it is used to do a multithreaded file write in the `def new_write(b)` function. 

Once again, thank you for your fast response! 

```
import sys
from contextlib import contextmanager
from io import StringIO

import numpy as np
import pandas as pd
import datetime

import streamlit as st
**from streamlit.report_thread import REPORT_CONTEXT_ATTR_NAME**
from threading import current_thread
import altair as alt
import functools
import base64

from timeseries_generator import (
    Generator,
    HolidayFactor,
    LinearTrend,
    RandomFeatureFactor,
    WeekdayFactor,
    WhiteNoise,
)
from timeseries_generator.external_factors import (
    CountryGdpFactor,
    EUIndustryProductFactor,
)

sys.path.append("../..")

np.random.seed(42)

st.set_page_config(
    page_title="Awesome TS Generator", layout="wide", initial_sidebar_state="auto"
)


@contextmanager
def st_redirect(src, dst):
    placeholder = st.empty()
    output_func = getattr(placeholder, dst)

    with StringIO() as buffer:
        old_write = src.write

        def new_write(b):
            if getattr(current_thread(), **REPORT_CONTEXT_ATTR_NAME**, None):
                buffer.write(b)
                output_func(buffer.getvalue())
            else:
                old_write(b)

        try:
            src.write = new_write
            yield
        finally:
            src.write = old_write


@contextmanager
def st_stdout(dst):
    with st_redirect(sys.stdout, dst):
        yield


@contextmanager
def st_stderr(dst):
    with st_redirect(sys.stderr, dst):
        yield

# -----------------------------------
# Streamlit APP start from here

```

tconkling commented:
It looks like this is redirecting stdout and stderr output to the Streamlit frontend. One solution might be to just remove this logic and allow stdout/stderr to go to the console. If that's not workable, I think your best bet is to go to the author of the original Streamlit app and see if they can help adapt it. (I think the "proper" solution is likely to be fairly complex - more than we can support in the comments section of a PR :))
